java/ql/src/utils/modelgenerator/internal/CaptureTypeBasedSummaryModels.qll

```diff
@@ -0,0 +1,333 @@
+private import java
+private import semmle.code.java.Collections
+private import semmle.code.java.dataflow.internal.ContainerFlow
+private import CaptureModelsSpecific as Specific
+private import CaptureModels
+
+/**
+ * A type representing instantiations of class types
+ * that has a method which returns an iterator.
+ */
+private class IterableType extends Class {
+  private Type elementType;
+
+  IterableType() {
+    exists(Method m, Type return, GenericType t, Type et, int position | m.getDeclaringType() = t |
+      return = m.getReturnType() and
+      return.getName().matches("Iterator%") and
+      et = return.(ParameterizedType).getTypeArgument(0) and
+      t.getTypeParameter(position) = et and
+      instantiates(this, t, position, elementType)
+    )
+  }
+
+  /**
+   * Returns the iterator element type of `this`.
+   */
+  Type getElementType() { result = elementType }
+}
+
+/**
+ * Holds if type `bound` is an upper bound for type `t` or equal to `t`.
+ */
+private predicate isEffectivelyUpperBound(Type t, Type bound) {
+  t = bound or
+  t.(Wildcard).getUpperBound().getType() = bound
+}
+
+/**
+ * Holds if type `bound` is a lower bound for type `t` or equal to `t`.
+ */
+private predicate isEffectivelyLowerBound(Type t, Type bound) {
+  t = bound or
+  t.(Wildcard).getLowerBound().getType() = bound
+}
+
+/**
+ * Holds if `t` is a container like type of `tv` (eg. `List<T>`).
+ * Note that collections are covariant in their element type.
```

Comment:
No? In Java they aren't. Or is this something that you assume for some heuristic purpose?

Reply:
Ok, that claim is to strong.
At least the idea is that, if we have "relevant" type variable that serves as an upper bound for a given type, then we use that to eg. decide if the method reads from an implicit field or the implicit collection.
I will delete the comment.

Reply:
Not yet deleted.

Reply:
haha - no - you are right! I will try once more :-)